Sources/swift-parsing-benchmark/JSON.swift

```diff
@@ -4,16 +4,16 @@ import Parsing
 
 /// This benchmark shows how to create a naive JSON parser with combinators.
 ///
-/// It is mostly implemented according to the [spec](https://www.json.org/json-en.html) (we take a
-/// shortcut and use `Double.parser()`, which behaves accordingly).
+/// It is implemented according to the [spec](https://www.json.org/json-en.html).
 let jsonSuite = BenchmarkSuite(name: "JSON") { suite in
   #if swift(>=5.8)
     struct JSONValue: ParserPrinter {
       enum Output: Equatable {
         case array([Self])
         case boolean(Bool)
+        case float(Double)
+        case integer(Int)
         case null
-        case number(Double)
```

Comment:
Can you explain this change? I was under the impression that JSON/JavaScript only has the concept of a "number" and doesn't distinguish integer and float values.

Reply:
Hmm yes, true. JSON only has the concept of a _number_, which is a sequence of digits, optionally followed by a decimal point and a fractional part. Now, if we model this as just a `Double` in Swift, we lose information about the number. `1` and `1.0`, for example, becomes indistinguishable. By modeling numbers that _does_ contain a fractional part as a `Double`, and numbers that doesn't as an `Int`, we at least keep that information, which I think is often desirable in Swift.

To not lose _any_ information we could model the numbers as a `String` (`case number(String)`), but that seems less convenient in practice. I think just separating numbers with and without any fractional part strikes a good balance of what information to keep about the numbers. This is at least how I'm thinking about this :)